### PR TITLE
Move Assertions to Hamcrest Matchers

### PR DESCRIPTION
**Pull request by thelooter**
# Description
Tracks the Process of moving all our Assertions to Hamcrest Matchers 

# Checklist
The following items should be checked before the pull request can be merged.
- [ ] Code follows existing style.
- [ ] Unit tests added (if applicable).

### CODE REVIEW

**Comment by Thorinwasher** [comment]
There we go, I have now added hamcrest matchers for all relevant assertions I could find. Some things to be noted on what I have done:
- Every class which has an assert method in it and is in src/main have been depricated and marked for removal
- I had to create some new api methods in some places to allow the same functionallity as previously

Other than that and the huge scale of this, I don't have anything more to note

File: src/main/java/org/mockbukkit/mockbukkit/matcher/command/MessageTargetReceivedAnyMessage.java
Comment (thelooter) [inline src/main/java/org/mockbukkit/mockbukkit/matcher/command/MessageTargetReceivedAnyMessage.java:1]: This file naming should probably follow the  `^.*Matcher$` Regex the other ones do too

File: src/main/java/org/mockbukkit/mockbukkit/matcher/entity/EntityLocationMatcher.java
Comment (thelooter) [inline src/main/java/org/mockbukkit/mockbukkit/matcher/entity/EntityLocationMatcher.java:1]: Is there a reason this is a Matcher? It seems like this can easily be achieved with an equality check on the Entities Location

File: src/main/java/org/mockbukkit/mockbukkit/matcher/entity/human/HumanEntityGameModeMatcher.java
Comment (thelooter) [inline src/main/java/org/mockbukkit/mockbukkit/matcher/entity/human/HumanEntityGameModeMatcher.java:1]: Same Question as in EntityLocationMatcher. Can this be achieved by a simple equality check instead of a Matcher?

File: src/main/java/org/mockbukkit/mockbukkit/simulate/entity/PlayerSimulation.java [hunk 1]
```diff
@@ -104,12 +104,12 @@ public PlayerSimulation(PlayerMock playerMock)
...
 	 * Simulates the player damaging a block. Note that this method does not anything unless the player is in survival
 	 * mode. If {@code InstaBreak} is set to true by an event handler, a {@link BlockBreakEvent} is immediately fired.
-	 * The result will then still be whether or not the {@link BlockDamageEvent} was cancelled or not, not the later
+	 * The result will then still be whether the {@link BlockDamageEvent} was cancelled or not, not the later
 	 * {@link BlockBreakEvent}.
 	 *
 	 * @param block The block to damage.
 	 * @return the event that was fired, {@code null} if the player was not in
-	 * survival gamemode.
+	 * survival game mode.
```
Comment (thelooter) [inline src/main/java/org/mockbukkit/mockbukkit/simulate/entity/PlayerSimulation.java:112]: I don't think this is a good change. Gamemode is an established term that I don't think needs to be broken up

File: src/test/java/org/mockbukkit/mockbukkit/inventory/meta/ItemMetaMockTest.java [hunk 2]
```diff
@@ -34,13 +34,19 @@
 import java.util.Set;
 import java.util.UUID;
 
+import static org.hamcrest.CoreMatchers.both;
```
Comment (thelooter) [inline src/test/java/org/mockbukkit/mockbukkit/inventory/meta/ItemMetaMockTest.java]: Am I overlooking something, or is this import not used?

File: src/test/java/org/mockbukkit/mockbukkit/matcher/plugin/AnEvent.java
Comment (thelooter) [inline src/test/java/org/mockbukkit/mockbukkit/matcher/plugin/AnEvent.java:1]: Why don't we use a premade event?

**Review by thelooter — COMMENTED** [review]
Some nitpicks

File: src/main/java/org/mockbukkit/mockbukkit/simulate/entity/PlayerSimulation.java (hunk 1 above)
Comment (Thorinwasher) [reply src/main/java/org/mockbukkit/mockbukkit/simulate/entity/PlayerSimulation.java:112]: Okay, there's defnitely inconsistencies here in that case, because the bukkit api defines gamemode as the enum `org.bukkit.GameMode`, which implies that it is two separate words

File: src/main/java/org/mockbukkit/mockbukkit/matcher/entity/EntityLocationMatcher.java
Comment (Thorinwasher) [reply src/main/java/org/mockbukkit/mockbukkit/matcher/entity/EntityLocationMatcher.java:1]: The main reason to be real, is that I simply looked at every method in every mock that includes the string assert and replaced it with a matcher. Except for that, I believe that there's some minute utility with this, as you can specify a radius around the destination that gives a valid match (which does require some thought to be put in with the normal assertion check)

Anyhow, I agree that it is not that important to include in this pull request.

File: src/test/java/org/mockbukkit/mockbukkit/matcher/plugin/AnEvent.java
Comment (Thorinwasher) [reply src/test/java/org/mockbukkit/mockbukkit/matcher/plugin/AnEvent.java:1]: Less maintanance required; The constructors of events are not always backwards compatible. I therefore just made an event which would not change constructor.

File: src/main/java/org/mockbukkit/mockbukkit/matcher/entity/human/HumanEntityGameModeMatcher.java
Comment (Thorinwasher) [reply src/main/java/org/mockbukkit/mockbukkit/matcher/entity/human/HumanEntityGameModeMatcher.java:1]: Yes, it can. Should I remove this one?

File: src/main/java/org/mockbukkit/mockbukkit/matcher/entity/EntityLocationMatcher.java
Comment (thelooter) [reply src/main/java/org/mockbukkit/mockbukkit/matcher/entity/EntityLocationMatcher.java:1]: In that case, I'd keep it, since it has some minute use

File: src/main/java/org/mockbukkit/mockbukkit/matcher/entity/human/HumanEntityGameModeMatcher.java
Comment (thelooter) [reply src/main/java/org/mockbukkit/mockbukkit/matcher/entity/human/HumanEntityGameModeMatcher.java:1]: In that case I don't really see the usefulness. I'd remove it

File: .github/workflows/gradle.yml
Comment (4everTheOne) [inline .github/workflows/gradle.yml:1]: Will this replace the current `.github/workflows/build.yml` and `.github/workflows/deploy.yml`

File: extra/TestPlugin/src/main/java/org/mockbukkit/testplugin/TestPlugin.java
Comment (4everTheOne) [inline extra/TestPlugin/src/main/java/org/mockbukkit/testplugin/TestPlugin.java:1]: Should we write a "migration guide" containing an explanation for this package rename?

Something like:
> Starting from version 4.0, MockBukkit packages were renamed from "be.seeseemelk.*" to "org.mockbukkit.*".
> Example:
> ```
> be.seeseemelk.mockbukkit.MockBukkit => org.mockbukkit.mockbukkit.MockBukkit;
> be.seeseemelk.mockbukkit.ServerMock => org.mockbukkit.mockbukkit.ServerMock;
> ```
> A simple migration guide would be to replace imports that contain `be.seeseemelk.mockbukkit` with `org.mockbukkit.mockbukkit`.
>

Additionally, I have compiled a list of classes that were renamed:
| v3                                                        |  v4                                                              |
| --------------------------------------------------------- | ---------------------------------------------------------------- |
| be.seeseemelk.mockbukkit.block.data.AmethystClusterMock   |  org.mockbukkit.mockbukkit.block.data.AmethystClusterDataMock    |
| be.seeseemelk.mockbukkit.block.data.BedMock               |  org.mockbukkit.mockbukkit.block.data.BedDataMock                |
| be.seeseemelk.mockbukkit.block.data.CampfireMock          |  org.mockbukkit.mockbukkit.block.data.CampfireDataMock           |
| be.seeseemelk.mockbukkit.block.data.SlabMock              |  org.mockbukkit.mockbukkit.block.data.SlabDataMock               |
| be.seeseemelk.mockbukkit.block.data.StairsMock            |  org.mockbukkit.mockbukkit.block.data.StairsDataMock             |
| be.seeseemelk.mockbukkit.block.data.SwitchMock            |  org.mockbukkit.mockbukkit.block.data.SwitchDataMock             |
| be.seeseemelk.mockbukkit.block.data.TrapDoorMock          |  org.mockbukkit.mockbukkit.block.data.TrapDoorDataMock           |
| be.seeseemelk.mockbukkit.block.data.WallSignMock          |  org.mockbukkit.mockbukkit.block.data.WallSignDataMock           |
| be.seeseemelk.mockbukkit.block.state.AbstractFurnaceMock  |  org.mockbukkit.mockbukkit.block.state.AbstractFurnaceStateMock  |
| be.seeseemelk.mockbukkit.block.state.BannerMock           |  org.mockbukkit.mockbukkit.block.state.BannerStateMock           |
| be.seeseemelk.mockbukkit.block.state.BarrelMock           |  org.mockbukkit.mockbukkit.block.state.BarrelStateMock           |
| be.seeseemelk.mockbukkit.block.state.BeaconMock           |  org.mockbukkit.mockbukkit.block.state.BeaconStateMock           |
| be.seeseemelk.mockbukkit.block.state.BedMock              |  org.mockbukkit.mockbukkit.block.state.BedStateMock              |
| be.seeseemelk.mockbukkit.block.state.BeehiveMock          |  org.mockbukkit.mockbukkit.block.state.BeehiveStateMock          |
| be.seeseemelk.mockbukkit.block.state.BellMock             |  org.mockbukkit.mockbukkit.block.state.BellStateMock             |
| be.seeseemelk.mockbukkit.block.state.BlastFurnaceMock     |  org.mockbukkit.mockbukkit.block.state.BlastFurnaceStateMock     |
| be.seeseemelk.mockbukkit.block.state.BrewingStandMock     |  org.mockbukkit.mockbukkit.block.state.BrewingStandStateMock     |
| be.seeseemelk.mockbukkit.block.state.CampfireMock         |  org.mockbukkit.mockbukkit.block.state.CampfireStateMock         |
| be.seeseemelk.mockbukkit.block.state.ChestMock            |  org.mockbukkit.mockbukkit.block.state.ChestStateMock            |
| be.seeseemelk.mockbukkit.block.state.CommandBlockMock     |  org.mockbukkit.mockbukkit.block.state.CommandBlockStateMock     |
| be.seeseemelk.mockbukkit.block.state.ComparatorMock       |  org.mockbukkit.mockbukkit.block.state.CompatatorStateMock       |
| be.seeseemelk.mockbukkit.block.state.ConduitMock          |  org.mockbukkit.mockbukkit.block.state.ConduitStateMock          |
| be.seeseemelk.mockbukkit.block.state.ContainerMock        |  org.mockbukkit.mockbukkit.block.state.ContainerStateMock        |
| be.seeseemelk.mockbukkit.block.state.CreatureSpawnerMock  |  org.mockbukkit.mockbukkit.block.state.CreatureSpawnerStateMock  |
| be.seeseemelk.mockbukkit.block.state.DaylightDetectorMock |  org.mockbukkit.mockbukkit.block.state.DaylightDetectorStateMock |
| be.seeseemelk.mockbukkit.block.state.DispenserMock        |  org.mockbukkit.mockbukkit.block.state.DispenserStateMock        |
| be.seeseemelk.mockbukkit.block.state.DropperMock          |  org.mockbukkit.mockbukkit.block.state.DropperStateMock          |
| be.seeseemelk.mockbukkit.block.state.EnchantingTableMock  |  org.mockbukkit.mockbukkit.block.state.EnchantingTableStateMock  |
| be.seeseemelk.mockbukkit.block.state.EnderChestMock       |  org.mockbukkit.mockbukkit.block.state.EnderChestStateMock       |
| be.seeseemelk.mockbukkit.block.state.EndGatewayMock       |  org.mockbukkit.mockbukkit.block.state.EndGatewayStateMock       |
| be.seeseemelk.mockbukkit.block.state.HopperMock           |  org.mockbukkit.mockbukkit.block.state.HopperStateMock           |
| be.seeseemelk.mockbukkit.block.state.JigsawMock           |  org.mockbukkit.mockbukkit.block.state.JigsawStateMock           |
| be.seeseemelk.mockbukkit.block.state.LecternMock          |  org.mockbukkit.mockbukkit.block.state.LecternStateMock          |
| be.seeseemelk.mockbukkit.block.state.SculkCatalystMock    |  org.mockbukkit.mockbukkit.block.state.SculkCatalystStateMock    |
| be.seeseemelk.mockbukkit.block.state.SculkSensorMock      |  org.mockbukkit.mockbukkit.block.state.SculkSensorStateMock      |
| be.seeseemelk.mockbukkit.block.state.SculkShriekerMock    |  org.mockbukkit.mockbukkit.block.state.SculkShriekerStateMock    |
| be.seeseemelk.mockbukkit.block.state.ShulkerBoxMock       |  org.mockbukkit.mockbukkit.block.state.ShulkerBoxStateMock       |
| be.seeseemelk.mockbukkit.block.state.SignMock             |  org.mockbukkit.mockbukkit.block.state.SignStateMock             |
| be.seeseemelk.mockbukkit.block.state.SkullMock            |  org.mockbukkit.mockbukkit.block.state.SkullStateMock            |
| be.seeseemelk.mockbukkit.block.state.SmokerMock           |  org.mockbukkit.mockbukkit.block.state.SmokerStateMock           |
| be.seeseemelk.mockbukkit.block.state.StructureMock        |  org.mockbukkit.mockbukkit.block.state.StructuteStateMock        |
| be.seeseemelk.mockbukkit.command.MockCommandMap           |  org.mockbukkit.mockbukkit.command.commandMapMock                |
| be.seeseemelk.mockbukkit.MockChunkData                    |  org.mockbukkit.mockbukkit.ChunkDataMock                         |
| be.seeseemelk.mockbukkit.MockPlayerList                   |  org.mockbukkit.mockbukkit.PlayerListMock                        |
| be.seeseemelk.mockbukkit.MockUnsafeValues                 |  org.mockbukkit.mockbukkit.UnsafeValuesMock                      |

File: src/main/java/org/mockbukkit/mockbukkit/BanListMock.java
Comment (4everTheOne) [inline src/main/java/org/mockbukkit/mockbukkit/BanListMock.java:1]: Is this file needed? it seems to be empty.

File: src/main/java/org/mockbukkit/mockbukkit/block/state/AbstractFurnaceStateMock.java [hunk 3]
```diff
@@ -16,9 +17,9 @@
 /**
  * Mock implementation of a {@link Furnace}.
  *
- * @see ContainerMock
+ * @see ContainerStateMock
  */
-public abstract class AbstractFurnaceMock extends ContainerMock implements Furnace
+public abstract class AbstractFurnaceStateMock extends org.mockbukkit.mockbukkit.block.state.ContainerStateMock implements Furnace
```
Comment (4everTheOne) [inline src/main/java/org/mockbukkit/mockbukkit/block/state/AbstractFurnaceStateMock.java:22]: We could replace the `org.mockbukkit.mockbukkit.block.state.ContainerStateMock` with `ContainerStateMock`

File: src/main/java/org/mockbukkit/mockbukkit/ServerMock.java [hunk 4]
```diff
@@ -566,7 +566,7 @@ public boolean removeWorld(WorldMock world)
 	@Override
 	public @NotNull String getBukkitVersion()
 	{
-		return BuildParameters.PAPER_API_FULL_VERSION;
+		return org.mockbukkit.mockbukkit.BuildParameters.PAPER_API_FULL_VERSION;
```
Comment (4everTheOne) [inline src/main/java/org/mockbukkit/mockbukkit/ServerMock.java:569]: Do we need this prefix?

File: src/main/java/org/mockbukkit/mockbukkit/matcher/block/BlockMaterialTypeMatcher.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import org.jetbrains.annotations.NotNull;
+import org.mockbukkit.mockbukkit.block.BlockMock;
+import org.mockbukkit.mockbukkit.command.CommandResult;
+
+public class BlockMaterialTypeMatcher extends TypeSafeMatcher<BlockMock>
+{
+
+	private final Material material;
+
+	public BlockMaterialTypeMatcher(Material material)
```
Comment (4everTheOne) [inline src/main/java/org/mockbukkit/mockbukkit/matcher/block/BlockMaterialTypeMatcher.java:16]: Should we set this visibility to protected rather than public?

File: src/main/java/org/mockbukkit/mockbukkit/matcher/command/CommandResultAnyResponseMatcher.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
+package org.mockbukkit.mockbukkit.matcher.command;
+
+import org.hamcrest.Description;
+import org.hamcrest.TypeSafeMatcher;
+import org.jetbrains.annotations.NotNull;
+import org.mockbukkit.mockbukkit.command.CommandResult;
+
+public class CommandResultAnyResponseMatcher extends TypeSafeMatcher<CommandResult>
```
Comment (4everTheOne) [inline src/main/java/org/mockbukkit/mockbukkit/matcher/command/CommandResultAnyResponseMatcher.java:8]: Since this implementation is similar to "CommandResultResponseMatcher" should this extend it?

File: src/main/java/org/mockbukkit/mockbukkit/matcher/entity/EntityLocationMatcher.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
+package org.mockbukkit.mockbukkit.matcher.entity;
+
+import com.google.common.base.Preconditions;
+import org.bukkit.Location;
+import org.hamcrest.Description;
+import org.hamcrest.TypeSafeMatcher;
+import org.jetbrains.annotations.NotNull;
+import org.mockbukkit.mockbukkit.entity.EntityMock;
+
+public class EntityLocationMatcher extends TypeSafeMatcher<EntityMock>
```
Comment (4everTheOne) [inline src/main/java/org/mockbukkit/mockbukkit/matcher/entity/EntityLocationMatcher.java:10]: Should we update this to be "Entity" rather than "EntityMock"?

File: src/main/java/org/mockbukkit/mockbukkit/matcher/entity/EntityTeleportationMatcher.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
+package org.mockbukkit.mockbukkit.matcher.entity;
+
+import com.google.common.base.Preconditions;
+import org.bukkit.Location;
+import org.hamcrest.Description;
+import org.hamcrest.TypeSafeMatcher;
+import org.jetbrains.annotations.NotNull;
+import org.mockbukkit.mockbukkit.entity.EntityMock;
+
+public class EntityTeleportationMatcher extends TypeSafeMatcher<EntityMock>
```
Comment (4everTheOne) [inline src/main/java/org/mockbukkit/mockbukkit/matcher/entity/EntityTeleportationMatcher.java:10]: Should this be "Entity" rather than "EntityMock"?

File: src/main/java/org/mockbukkit/mockbukkit/matcher/entity/human/HumanEntityInventoryViewItemMatcher.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	{
+		return new HumanEntityInventoryViewItemMatcher(material);
+	}
+
+	/**
+	 *
+	 * @param item The item required for a match
+	 * @return A matcher which matches with any human entity with the specified item in inventory
+	 */
+	public static @NotNull HumanEntityInventoryViewItemMatcher hasItemInInventoryView(@NotNull ItemStack item)
```
Comment (4everTheOne) [inline src/main/java/org/mockbukkit/mockbukkit/matcher/entity/human/HumanEntityInventoryViewItemMatcher.java:65]: Should we take into consideration the quantity?

**Review by 4everTheOne — COMMENTED** [review]
Partial revision

File: src/test/java/org/mockbukkit/mockbukkit/ThreadAccessExceptionTest.java
Comment (4everTheOne) [inline src/test/java/org/mockbukkit/mockbukkit/ThreadAccessExceptionTest.java:1]: This file is empty, was this intentional?

File: src/main/java/org/mockbukkit/mockbukkit/BanListMock.java
Comment (Thorinwasher) [reply src/main/java/org/mockbukkit/mockbukkit/BanListMock.java:1]: I think this is an artefact from that the main 4.0 branch got rebased: We probably need to do a rebase on this pull request as well. I have no clue how to do this except through pain and missery

File: .github/workflows/gradle.yml
Comment (Thorinwasher) [reply .github/workflows/gradle.yml:1]: I have no clue why that got added at least into this pr.

File: extra/TestPlugin/src/main/java/org/mockbukkit/testplugin/TestPlugin.java
Comment (Thorinwasher) [reply extra/TestPlugin/src/main/java/org/mockbukkit/testplugin/TestPlugin.java:1]: This is definitely a good idea!

**Comment by Thorinwasher** [comment]
@4everTheOne I just realized it's exceedingly difficult to review this pr in the state it is right now. The pr has so many changes in it because the base 4.0 branch got rebased, which means that aproximately every file in the whole project is in this exact pr. (Just from the namespace change).

I know too little about rebasing to fix this without getting serious headaches. I think we need to discuss this in the meeting.

(If I rememeber correctly this pr really is only around 70 file changes)

File: src/main/java/org/mockbukkit/mockbukkit/matcher/entity/human/HumanEntityInventoryViewItemMatcher.java (hunk 9 above)
Comment (Thorinwasher) [reply src/main/java/org/mockbukkit/mockbukkit/matcher/entity/human/HumanEntityInventoryViewItemMatcher.java:65]: We could deffinitely do that.

File: src/main/java/org/mockbukkit/mockbukkit/block/state/AbstractFurnaceStateMock.java (hunk 3 above)
Comment (Thorinwasher) [reply src/main/java/org/mockbukkit/mockbukkit/block/state/AbstractFurnaceStateMock.java:22]: Good point there. definitely needs to change

File: src/main/java/org/mockbukkit/mockbukkit/ServerMock.java (hunk 4 above)
Comment (Thorinwasher) [reply src/main/java/org/mockbukkit/mockbukkit/ServerMock.java:569]: Nope, probably an artefact from a refactor

File: src/main/java/org/mockbukkit/mockbukkit/matcher/block/BlockMaterialTypeMatcher.java (hunk 5 above)
Comment (Thorinwasher) [reply src/main/java/org/mockbukkit/mockbukkit/matcher/block/BlockMaterialTypeMatcher.java:16]: Yes, should do

File: src/main/java/org/mockbukkit/mockbukkit/matcher/command/CommandResultAnyResponseMatcher.java (hunk 6 above)
Comment (Thorinwasher) [reply src/main/java/org/mockbukkit/mockbukkit/matcher/command/CommandResultAnyResponseMatcher.java:8]: That could definitely be done. I don't think that much thought went into this. But generally I believe you should avoid somewhat unnecessary abstraction. Like it's ok to repeat yourself a little

File: src/main/java/org/mockbukkit/mockbukkit/matcher/entity/EntityLocationMatcher.java (hunk 7 above)
Comment (Thorinwasher) [reply src/main/java/org/mockbukkit/mockbukkit/matcher/entity/EntityLocationMatcher.java:10]: Yep, good idea

File: src/test/java/org/mockbukkit/mockbukkit/ThreadAccessExceptionTest.java
Comment (Thorinwasher) [reply src/test/java/org/mockbukkit/mockbukkit/ThreadAccessExceptionTest.java:1]: Artifact for 4.0 main branch being rebased, I believe

File: src/main/java/org/mockbukkit/mockbukkit/matcher/block/BlockMaterialTypeMatcher.java (hunk 5 above)
Comment (Thorinwasher) [reply src/main/java/org/mockbukkit/mockbukkit/matcher/block/BlockMaterialTypeMatcher.java:16]: Yes, why not. It's a good idea to not expose this to our api

**Comment by thelooter** [comment]
Superseded by #1156
